### ISSUE TITLE
Use index_copy_ to reduce memory copies

### ISSUE THREAD

**Issue by mzchtx** [Lightning-AI/lit-llama#382]
I think we can use index_copy_` the inplace version of index_copy` to reduce the extra cache creation and copying

https://github.com/Lightning-AI/lit-llama/blob/main/lit_llama/model.py#L217-L218

![image](https://github.com/Lightning-AI/lit-llama/assets/37564754/34001d79-ccea-4bc2-89ea-d947c10a0ebe)

**Comment by carmocca** [comment]
@gkroiz Could this change be detrimental to XLA's performance?

**Comment by carmocca** [comment]
@mzchtx Did you measure the performance difference?

Would you like to open a PR with your suggestion?

**Comment by mzchtx** [comment]
> 

Yes, however, this primarily impacts inference performance. During our testing, we observed a performance degradation of around 10-20% with larger batch sizes, specifically when using batch size of 256 and 512. GPT-style model inference is often IO-bound, so it's better to minimize unnecessary memory copies.

**Comment by mzchtx** [comment]
> @mzchtx Did you measure the performance difference?
> 
> Would you like to open a PR with your suggestion?


Yes, however, this primarily impacts inference performance. During our testing, we observed a performance degradation of around 10-20% with larger batch sizes, specifically when using batch size of 256 and 512. GPT-style model inference is often IO-bound, so it's better to minimize unnecessary memory copies.
